training/worker.py

```diff
@@ -13,21 +13,26 @@
 import service_pb2
 import service_pb2_grpc
 
+from threading import Lock
 
 from timemachine.lib import custom_ops, ops
 
 class Worker(service_pb2_grpc.WorkerServicer):
 
     def __init__(self):
-        self.state = None
+        self.states = {}
+        self.mutex = Lock()
 
     def ResetState(self, request, context):
-        self.state = None
+        self.mutex.acquire()
+        self.states = {}
+        self.mutex.release()
         reply = service_pb2.EmptyMessage()
         return reply
 
     def ForwardMode(self, request, context):
-        assert self.state is None
+        # assert self.state is None
+        assert request.key not in self.states
```

Comment:
This should be inside the mutex

Reply:
good call

Reply:
I'm actually going to ditch this assertion entirely